File: gateway/auth/go/fixtures_test.go
package macaroon_test

// Cross-language byte-equivalence test (Go side).
//
// Loads every gateway/auth/fixtures/*.json, asserts that this Go
// package reproduces every deterministic intermediate value the TS
// package recorded, and verifies the assembled macaroon end-to-end.
//
// We don't re-sign on the Go side (signing is the TS package's job in
// production; the Go side is pure verifier). What we DO assert,
// byte-for-byte:
//
//   1. JCS of (ua minus org_sig) → expected.ua_signing_bytes_hex
//   2. JCS of (inv minus user_sig) → expected.inv_signing_bytes_hex
//   3. JCS of (ua minus org_sig) as string → expected.ua_canonical_json
//   4. JCS of (inv minus user_sig) as string → expected.inv_canonical_json
//   5. For each attenuation: JCS of caveats, HMAC input, HMAC output
//      → expected.attenuation_hmac_inputs[i].*
//   6. JCS of the full macaroon → expected.macaroon_canonical_json
//   7. base64url of (6) → expected.macaroon_b64url
//   8. Verify(macaroon_b64url, policy, now) → expected.claims
//
// If TS and Go agree on these, they agree on the wire format.

import (
	"encoding/json"
	"os"
	"path/filepath"
	"reflect"
	"sort"
	"testing"
	"time"

	macaroon "github.com/stakwork/stakgraph/gateway/auth/go"
)

const fixturesRelPath = "../fixtures"

// fixtureFile is the on-disk shape of one fixture. Mirrors the TS
// regenerate-fixtures.ts script's output exactly.
type fixtureFile struct {
	Description string `json:"description"`
	Inputs      struct {
		OrgID       string             `json:"org_id"`
		OrgPrivHex  string             `json:"org_priv_hex,omitempty"`
		OrgPrivsHex map[string]string  `json:"org_privs_hex,omitempty"`
		Signers     []int              `json:"participating_signers,omitempty"`
		UserPrivHex string             `json:"user_priv_hex"`
		Policy      macaroon.Policy    `json:"policy"`
		UAUnsigned  json.RawMessage    `json:"ua_unsigned"`
		InvUnsigned json.RawMessage    `json:"inv_unsigned"`
		AttsUnsigned []json.RawMessage `json:"atts_unsigned"`
	} `json:"inputs"`
	Expected struct {
		UASigningBytesHex      string `json:"ua_signing_bytes_hex"`
		InvSigningBytesHex     string `json:"inv_signing_bytes_hex"`
		UACanonicalJSON        string `json:"ua_canonical_json"`
		InvCanonicalJSON       string `json:"inv_canonical_json"`
		MacaroonCanonicalJSON  string `json:"macaroon_canonical_json"`
		MacaroonB64url         string `json:"macaroon_b64url"`
		AttenuationHmacInputs  []struct {
			PrevSigHex            string `json:"prev_sig_hex"`
			CaveatsCanonicalJSON  string `json:"caveats_canonical_json"`
			HmacInputHex          string `json:"hmac_input_hex"`
			HmacOutputHex         string `json:"hmac_output_hex"`
		} `json:"attenuation_hmac_inputs"`
		Claims struct {
			OrgID            string   `json:"org_id"`
			UserID           string   `json:"user_id"`
			Workspace        string   `json:"workspace"`
			AgentName        string   `json:"agent_name"`
			RunID            string   `json:"run_id"`
			EffectiveCaveats struct {
				Agents     []string `json:"agents"`
				MaxCostUSD float64  `json:"max_cost_usd"`
				MaxSteps   int      `json:"max_steps"`
				Exp        string   `json:"exp"`
			} `json:"effective_caveats"`
			Nonces []string `json:"nonces"`
			IAT    string   `json:"iat"`
		} `json:"claims"`
		VerifyResult string `json:"verify_result"`
	} `json:"expected"`
}

func listFixtures(t *testing.T) []string {
	entries, err := os.ReadDir(fixturesRelPath)
	if err != nil {
		t.Fatalf("read fixtures dir: %v", err)
	}
	var out []string
	for _, e := range entries {
		n := e.Name()
		if len(n) >= 7 && n[2] == '-' && filepath.Ext(n) == ".json" {
			out = append(out, n)
		}
	}
	sort.Strings(out)
	return out
}

func loadFixture(t *testing.T, name string) *fixtureFile {
	raw, err := os.ReadFile(filepath.Join(fixturesRelPath, name))
	if err != nil {
		t.Fatalf("read fixture %s: %v", name, err)
	}
	var fx fixtureFile
	if err := json.Unmarshal(raw, &fx); err != nil {
		t.Fatalf("unmarshal fixture %s: %v", name, err)
	}
	return &fx
}

func TestFixtures(t *testing.T) {
	files := listFixtures(t)
	if len(files) == 0 {
		t.Fatal("no fixture files found")
	}
	for _, f := range files {
		f := f
		t.Run(f, func(t *testing.T) {
			runFixture(t, f)
		})
	}
}

func runFixture(t *testing.T, name string) {
	fx := loadFixture(t, name)

	// ─── 1+3. JCS of (ua minus org_sig) ──────────────────────────
	uaSignBytes, err := jcsStripFromRaw(fx.Inputs.UAUnsigned, "")
	// ua_unsigned does not contain org_sig — JCS it directly.
	if err != nil {
		t.Fatalf("jcs ua_unsigned: %v", err)
	}
	if got := macaroon.BytesToHex(uaSignBytes); got != fx.Expected.UASigningBytesHex {
		t.Errorf("ua_signing_bytes_hex mismatch\n  got:  %s\n  want: %s", got, fx.Expected.UASigningBytesHex)
	}
	if got := string(uaSignBytes); got != fx.Expected.UACanonicalJSON {
		t.Errorf("ua_canonical_json mismatch\n  got:  %s\n  want: %s", got, fx.Expected.UACanonicalJSON)
	}

	// ─── 2+4. JCS of (inv minus user_sig) ────────────────────────
	invSignBytes, err := jcsStripFromRaw(fx.Inputs.InvUnsigned, "")
	if err != nil {
		t.Fatalf("jcs inv_unsigned: %v", err)
	}
	if got := macaroon.BytesToHex(invSignBytes); got != fx.Expected.InvSigningBytesHex {
		t.Errorf("inv_signing_bytes_hex mismatch\n  got:  %s\n  want: %s", got, fx.Expected.InvSigningBytesHex)
	}
	if got := string(invSignBytes); got != fx.Expected.InvCanonicalJSON {
		t.Errorf("inv_canonical_json mismatch\n  got:  %s\n  want: %s", got, fx.Expected.InvCanonicalJSON)
	}

	// ─── 5. attenuation HMAC inputs ──────────────────────────────
	// To reproduce these we need the signed macaroon's user_sig.sig
	// (for the first link's prev_sig) and each subsequent hmac. We
	// pull these from the assembled macaroon in expected.MacaroonB64url.
	signed, err := decodeMacaroonForTest(fx.Expected.MacaroonB64url)
	if err != nil {
		t.Fatalf("decode assembled macaroon: %v", err)
	}
	prevSig, err := macaroon.HexToBytes(signed.Invocation.UserSig.Sig)
	if err != nil {
		t.Fatalf("decode invocation.user_sig.sig: %v", err)
	}
	for i, att := range signed.Attenuations {
		exp := fx.Expected.AttenuationHmacInputs[i]
		if got := macaroon.BytesToHex(prevSig); got != exp.PrevSigHex {
			t.Errorf("att[%d].prev_sig_hex mismatch\n  got:  %s\n  want: %s", i, got, exp.PrevSigHex)
		}
		caveatsCanonical, err := macaroon.JCS(att.Caveats)
		if err != nil {
			t.Fatalf("att[%d] jcs caveats: %v", i, err)
		}
		if got := string(caveatsCanonical); got != exp.CaveatsCanonicalJSON {
			t.Errorf("att[%d].caveats_canonical_json mismatch\n  got:  %s\n  want: %s", i, got, exp.CaveatsCanonicalJSON)
		}
		if got := macaroon.BytesToHex(caveatsCanonical); got != exp.HmacInputHex {
			t.Errorf("att[%d].hmac_input_hex mismatch\n  got:  %s\n  want: %s", i, got, exp.HmacInputHex)
		}
		hmacOut, err := macaroon.ComputeAttenuationHMAC(prevSig, att.Caveats)
		if err != nil {
			t.Fatalf("att[%d] hmac: %v", i, err)
		}
		if got := macaroon.BytesToHex(hmacOut); got != exp.HmacOutputHex {
			t.Errorf("att[%d].hmac_output_hex mismatch\n  got:  %s\n  want: %s", i, got, exp.HmacOutputHex)
		}
		prevSig, err = macaroon.HexToBytes(att.HMAC)
		if err != nil {
			t.Fatalf("att[%d] decode hmac: %v", i, err)
		}
	}

	// ─── 6+7. full macaroon JCS + base64url ──────────────────────
	macaroonCanonical, err := macaroon.JCS(signed)
	if err != nil {
		t.Fatalf("jcs full macaroon: %v", err)
	}
	if got := string(macaroonCanonical); got != fx.Expected.MacaroonCanonicalJSON {
		t.Errorf("macaroon_canonical_json mismatch\n  got:  %s\n  want: %s", got, fx.Expected.MacaroonCanonicalJSON)
	}
	if got := macaroon.BytesToBase64url(macaroonCanonical); got != fx.Expected.MacaroonB64url {
		t.Errorf("macaroon_b64url mismatch\n  got:  %s\n  want: %s", got, fx.Expected.MacaroonB64url)
	}

	// ─── 8. Verify end-to-end ────────────────────────────────────
	// Pick a "now" that's after invocation.iat but before exp. We use
	// invocation.iat + 1s, same as the TS test.
	invIat, err := time.Parse(time.RFC3339, signed.Invocation.IAT)
	if err != nil {
		t.Fatalf("parse invocation.iat: %v", err)
	}
	claims, err := macaroon.Verify(fx.Expected.MacaroonB64url, fx.Inputs.Policy, invIat.Add(time.Second))
	if err != nil {
		t.Fatalf("Verify failed: %v", err)
	}
	if claims.OrgID != fx.Expected.Claims.OrgID {
		t.Errorf("claims.OrgID: got %q want %q", claims.OrgID, fx.Expected.Claims.OrgID)
	}
	if claims.UserID != fx.Expected.Claims.UserID {
		t.Errorf("claims.UserID: got %q want %q", claims.UserID, fx.Expected.Claims.UserID)
	}
	if claims.Workspace != fx.Expected.Claims.Workspace {
		t.Errorf("claims.Workspace: got %q want %q", claims.Workspace, fx.Expected.Claims.Workspace)
	}
	if claims.AgentName != fx.Expected.Claims.AgentName {
		t.Errorf("claims.AgentName: got %q want %q", claims.AgentName, fx.Expected.Claims.AgentName)
	}
	if claims.RunID != fx.Expected.Claims.RunID {
		t.Errorf("claims.RunID: got %q want %q", claims.RunID, fx.Expected.Claims.RunID)
	}
	if claims.IAT != fx.Expected.Claims.IAT {
		t.Errorf("claims.IAT: got %q want %q", claims.IAT, fx.Expected.Claims.IAT)
	}
	if !reflect.DeepEqual(claims.Nonces, fx.Expected.Claims.Nonces) {
		t.Errorf("claims.Nonces: got %v want %v", claims.Nonces, fx.Expected.Claims.Nonces)
	}
	if !reflect.DeepEqual(claims.EffectiveCaveats.Agents, fx.Expected.Claims.EffectiveCaveats.Agents) {
		t.Errorf("claims.EffectiveCaveats.Agents: got %v want %v",
			claims.EffectiveCaveats.Agents, fx.Expected.Claims.EffectiveCaveats.Agents)
	}
	if claims.EffectiveCaveats.MaxCostUSD != fx.Expected.Claims.EffectiveCaveats.MaxCostUSD {
		t.Errorf("claims.EffectiveCaveats.MaxCostUSD: got %v want %v",
			claims.EffectiveCaveats.MaxCostUSD, fx.Expected.Claims.EffectiveCaveats.MaxCostUSD)
	}
	if claims.EffectiveCaveats.MaxSteps != fx.Expected.Claims.EffectiveCaveats.MaxSteps {
		t.Errorf("claims.EffectiveCaveats.MaxSteps: got %v want %v",
			claims.EffectiveCaveats.MaxSteps, fx.Expected.Claims.EffectiveCaveats.MaxSteps)
	}
	if claims.EffectiveCaveats.Exp != fx.Expected.Claims.EffectiveCaveats.Exp {
		t.Errorf("claims.EffectiveCaveats.Exp: got %q want %q",
			claims.EffectiveCaveats.Exp, fx.Expected.Claims.EffectiveCaveats.Exp)
	}
}

// jcsStripFromRaw runs JCS on raw JSON. If field is non-empty, that
// top-level field is removed first. Used to JCS the unsigned
// ua/invocation objects from the fixture (which already lack their
// sig field, so we pass "").
func jcsStripFromRaw(raw json.RawMessage, field string) ([]byte, error) {
	var asMap map[string]json.RawMessage
	if err := json.Unmarshal(raw, &asMap); err != nil {
		return nil, err
	}
	if field != "" {
		delete(asMap, field)
	}
	repacked, err := json.Marshal(asMap)
	if err != nil {
		return nil, err
	}
	// Reuse the package's JCS by unmarshalling to any and calling JCS.
	var v any
	if err := json.Unmarshal(repacked, &v); err != nil {
		return nil, err
	}
	return macaroon.JCS(v)
}

// decodeMacaroonForTest base64url-decodes and JSON-parses a macaroon
// for test inspection (we need to read user_sig.sig and attenuation
// hmacs to drive the HMAC chain re-computation).
func decodeMacaroonForTest(b64 string) (*macaroon.Macaroon, error) {
	raw, err := macaroon.Base64urlToBytes(b64)
	if err != nil {
		return nil, err
	}
	var m macaroon.Macaroon
	if err := json.Unmarshal(raw, &m); err != nil {
		return nil, err
	}
	return &m, nil
}

File: gateway/auth/ts/test/fixtures.test.ts
/**
 * Cross-language byte-equivalence test (TS side).
 *
 * Loads every `gateway/auth/fixtures/*.json`, reproduces every
 * `expected.*` value from the `inputs.*` private keys + unsigned
 * objects, and fails if any byte differs.
 *
 * The Go side runs the same test against the same fixtures. If both
 * pass, the implementations agree.
 */

import { strict as assert } from "node:assert";
import { readdirSync, readFileSync } from "node:fs";
import { dirname, join } from "node:path";
import { test } from "node:test";
import { fileURLToPath } from "node:url";

import {
  attenuate,
  bytesToBase64url,
  bytesToHex,
  computeAttenuationHmac,
  hexToBytes,
  invocationSigBytes,
  jcs,
  signingBytes,
  signInvocation,
  signUserAuthorizationMultisig,
  signUserAuthorizationSingle,
  utf8Bytes,
  verify,
} from "../src/index.js";
import type {
  Attenuation,
  AttenuationCaveats,
  Budget,
  InvocationUnsigned,
  Macaroon,
  Policy,
  UserAuthorizationUnsigned,
} from "../src/index.js";

const HERE = dirname(fileURLToPath(import.meta.url));
const FIXTURES_DIR = join(HERE, "..", "..", "fixtures");

interface FixtureShape {
  description: string;
  inputs: {
    org_id: string;
    org_priv_hex?: string;                     // single-sig
    org_privs_hex?: Record<string, string>;    // multisig
    participating_signers?: number[];          // multisig
    user_priv_hex: string;
    policy: Policy;
    ua_unsigned: UserAuthorizationUnsigned;
    inv_unsigned: InvocationUnsigned;
    atts_unsigned: AttenuationCaveats[];
  };
  expected: {
    ua_signing_bytes_hex: string;
    inv_signing_bytes_hex: string;
    ua_canonical_json: string;
    inv_canonical_json: string;
    macaroon_canonical_json: string;
    macaroon_b64url: string;
    attenuation_hmac_inputs: Array<{
      prev_sig_hex: string;
      caveats_canonical_json: string;
      hmac_input_hex: string;
      hmac_output_hex: string;
    }>;
    claims: {
      org_id: string;
      user_id: string;
      agent_name: string;
      run_id: string;
      effective_caveats: {
        agents: string[];
        max_cost_usd: number;
        max_steps: number;
        budget: Budget | null;
        exp: string;
      };
      ua_nonce: string;
      ua_budget: Budget | null;
      permitted_realms: string[] | null;
      nonces: string[];
      iat: string;
    };
    verify_result: "ok";
  };
}

function fixtureFiles(): string[] {
  return readdirSync(FIXTURES_DIR)
    .filter((f) => /^\d{2}-.*\.json$/.test(f))
    .sort();
}

for (const file of fixtureFiles()) {
  test(`fixture ${file}`, () => {
    const fx: FixtureShape = JSON.parse(
      readFileSync(join(FIXTURES_DIR, file), "utf8"),
    );

    // ─── re-sign ua ──────────────────────────────────────────────
    let signedUa;
    if (fx.inputs.org_priv_hex) {
      signedUa = signUserAuthorizationSingle(
        fx.inputs.ua_unsigned,
        hexToBytes(fx.inputs.org_priv_hex),
      );
    } else if (fx.inputs.org_privs_hex && fx.inputs.participating_signers) {
      const signers = fx.inputs.participating_signers.map((idx) => ({
        key_index: idx,
        privKey: hexToBytes(fx.inputs.org_privs_hex![String(idx)]!),
      }));
      signedUa = signUserAuthorizationMultisig(fx.inputs.ua_unsigned, signers);
    } else {
      throw new Error(`fixture ${file}: no org signing inputs`);
    }

    // ─── re-sign invocation ──────────────────────────────────────
    const signedInv = signInvocation(
      fx.inputs.inv_unsigned,
      hexToBytes(fx.inputs.user_priv_hex),
    );

    // ─── signing bytes ───────────────────────────────────────────
    const uaMsg = signingBytes(
      fx.inputs.ua_unsigned as unknown as Record<string, unknown>,
      "org_sig",
    );
    const invMsg = signingBytes(
      fx.inputs.inv_unsigned as unknown as Record<string, unknown>,
      "user_sig",
    );
    assert.equal(bytesToHex(uaMsg), fx.expected.ua_signing_bytes_hex,
      "ua_signing_bytes_hex");
    assert.equal(bytesToHex(invMsg), fx.expected.inv_signing_bytes_hex,
      "inv_signing_bytes_hex");

    // ─── canonical JSON of signed layers ─────────────────────────
    assert.equal(
      jcs(stripField(signedUa, "org_sig")),
      fx.expected.ua_canonical_json,
      "ua_canonical_json",
    );
    assert.equal(
      jcs(stripField(signedInv, "user_sig")),
      fx.expected.inv_canonical_json,
      "inv_canonical_json",
    );

    // ─── attenuations ────────────────────────────────────────────
    const atts: Attenuation[] = [];
    let prevSig = invocationSigBytes(signedInv);
    fx.inputs.atts_unsigned.forEach((caveats, i) => {
      const exp = fx.expected.attenuation_hmac_inputs[i];
      assert.ok(exp, `expected attenuation_hmac_inputs[${i}] present`);

      assert.equal(bytesToHex(prevSig), exp.prev_sig_hex,
        `attenuation ${i} prev_sig_hex`);
      const caveatsCanonical = jcs(caveats as unknown as Record<string, unknown>);
      assert.equal(caveatsCanonical, exp.caveats_canonical_json,
        `attenuation ${i} caveats_canonical_json`);
      const hmacInput = utf8Bytes(caveatsCanonical);
      assert.equal(bytesToHex(hmacInput), exp.hmac_input_hex,
        `attenuation ${i} hmac_input_hex`);
      const hmacOutput = computeAttenuationHmac(prevSig, caveats);
      assert.equal(bytesToHex(hmacOutput), exp.hmac_output_hex,
        `attenuation ${i} hmac_output_hex`);

      atts.push(attenuate(prevSig, caveats));
      prevSig = hexToBytes(atts[i]!.hmac);
    });

    // ─── full macaroon ───────────────────────────────────────────
    const macaroon: Macaroon = {
      v: 1,
      org_id: fx.inputs.org_id,
      user_authorization: signedUa,
      invocation: signedInv,
      attenuations: atts,
    };
    const macaroonCanonical = jcs(macaroon as unknown as Record<string, unknown>);
    assert.equal(macaroonCanonical, fx.expected.macaroon_canonical_json,
      "macaroon_canonical_json");
    const macaroonB64 = bytesToBase64url(utf8Bytes(macaroonCanonical));
    assert.equal(macaroonB64, fx.expected.macaroon_b64url, "macaroon_b64url");

    // ─── round-trip verify ───────────────────────────────────────
    const claims = verify(
      macaroonB64,
      fx.inputs.policy,
      new Date(Date.parse(fx.inputs.inv_unsigned.iat) + 1000), // 1s after iat: in window
    );
    assert.equal(claims.org_id, fx.expected.claims.org_id);
    assert.equal(claims.user_id, fx.expected.claims.user_id);
    assert.equal(claims.agent_name, fx.expected.claims.agent_name);
    assert.equal(claims.run_id, fx.expected.claims.run_id);
    assert.deepEqual(claims.effective_caveats, fx.expected.claims.effective_caveats);
    assert.equal(claims.ua_nonce, fx.expected.claims.ua_nonce);
    assert.deepEqual(claims.ua_budget, fx.expected.claims.ua_budget);
    assert.deepEqual(claims.permitted_realms, fx.expected.claims.permitted_realms);
    assert.deepEqual(claims.nonces, fx.expected.claims.nonces);
    assert.equal(claims.iat, fx.expected.claims.iat);
  });
}

function stripField(obj: Record<string, unknown>, field: string): Record<string, unknown> {
  const out: Record<string, unknown> = {};
  for (const k of Object.keys(obj)) if (k !== field) out[k] = obj[k];
  return out;
}
